sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala

```diff
@@ -66,78 +72,339 @@ case class GpuShuffleCoalesceExec(child: SparkPlan, targetBatchByteSize: Long)
     val metricsMap = allMetrics
     val targetSize = targetBatchByteSize
     val dataTypes = GpuColumnVector.extractTypes(schema)
+    val readOption = CoalesceReadOption(new RapidsConf(conf))
 
     child.executeColumnar().mapPartitions { iter =>
-      new GpuShuffleCoalesceIterator(
-        new HostShuffleCoalesceIterator(iter, targetSize, metricsMap),
-        dataTypes, metricsMap)
+      GpuShuffleCoalesceUtils.getGpuShuffleCoalesceIterator(iter, targetSize, dataTypes,
+        readOption, metricsMap)
     }
   }
 }
 
+/** A case class to pack some options. Now it has only one, but may have more in the future */
+case class CoalesceReadOption private(useSplitRetryRead: Boolean)
+
+object CoalesceReadOption {
+  def apply(conf: RapidsConf): CoalesceReadOption =
+    CoalesceReadOption(conf.shuffleSplitRetryReadEnabled)
+}
+
+object GpuShuffleCoalesceUtils {
+  def getGpuShuffleCoalesceIterator(
+      iter: Iterator[ColumnarBatch],
+      targetSize: Long,
+      dataTypes: Array[DataType],
+      readOption: CoalesceReadOption,
+      metricsMap: Map[String, GpuMetric],
+      prefetchFirstBatch: Boolean = false): Iterator[ColumnarBatch] = {
+    if (readOption.useSplitRetryRead) {
+      val reader = new GpuShuffleCoalesceReader(iter, targetSize, dataTypes, metricsMap)
```

Comment:
I am confused why we want to put the coalesce on the GPU if we want to try and split and retry the read? We have it on the CPU to speed up the processing, but the new default puts it on the GPU. Do you have some performance numbers to show the impact of this change?

Reply:
Thx for review. And the coalesce is still on the CPU. What this new `GpuShuffleCoalesceReader` does is just a combination of `HostShuffleCoalesceIterator` and `GpuShuffleCoalesceIterator`, with the additional split-retry support when trying to move the coalcesced buffer to the GPU. 

Split-retry does not mean we will split the coalesced buffer. Instead we re-collect less shuffle buffers (total size <= target_size/2) from the cache, concatenate them and move to the GPU again. The new coalesced buffer is about half size of the original one, so requires less GPU memory when copying it to the GPU.

The change is big but the idea is quite simple. 

```
Pull in shuffle buffers according to the target size on CPU
and cache them
     |
     |
    \ /
Take the first N buffers with total size up to the target_size   <---
     |                                                               |
     |                                                               |
    \ /                                                              |
Do the coalcesce on CPU                                     (reduce the target_szie by half )
     |                                                               |                                                        
     |                                                               |
    \ /                                                              |
Move the coalcesced big buffer to GPU            ---- (GPU OOM) ---->
     |
     |
    \ /
Done, remove the first N buffers from the cache
```